### PR TITLE
feat: sdk7 UI components

### LINKED ISSUE #3520: Implement SDK7 UI Components
As part of the SDK7 implementation we need to create generic set of UI components described here: 
https://github.com/decentraland/adr/issues/125

The components should be based on the Yoga layout system, documentation can be found here: https://docs.unity3d.com/Manual/UIE-LayoutEngine.html

Acceptance Criteria:
Can render each of the following components from the SDK:
- [x] UiText (it was implemented before)
- [x] UiBackground
  - [x] UVs special processing (float array)
  - [x] Unit Tests
- [x] UiButton
- [x] UiInput
  - [x] Placeholder logic
  - [x] Unit Tests
- [x] UiDropdown
  - [x] Default Selection special case
  - [x] Unit Tests
- [x] Interactivity for UI
  - [x] Compatible events
  - [x] Unique events
  - [x] Unit tests
- [x] Generalization with existing components
- [x] Tidying up
- [x] Integration with SDK (debugging, testing, protocol adjustments)
  - [x] UIInput
  - [x] UIDropdown
  - [x] UIBackground
  - [x] Generic Events

### PR DESCRIPTION
**Pull request by mikhail-dcl**
Closes #3520 

### Generic UI Events

- `PointerHover` and everything that is connected to it was renamed to `PointerEvents`
- Support of such events (that already exist in 3D context) was added to UI elements in a generic way. Most of data that is irrelevant for 2D is omitted. Refer to `UIEventSubscription<TEvent>`, `UIPointerEventsHandler` and related extensions, the system from the 3D world is fully reused

### Custom UI Events

Every UI component can provide a feedback creating a custom component. It follows the existing pattern of "Internal" components. Refer to `InternalUIInputResults`, `UIPointerEventsUtils`: `RegisterFeedback`, `UnregisterFeedback`.
Then these results are utilized by `ECSUIInputSenderSystem`, and feedback components are sent back to the scene.

### Generalization

I generalized an approach of new components introduction to reuse as much code as possible within reasonable extents.

- `UIElementHandlerBase` to create a handler for a component
- `UIElementRegisterBase<T, THandler, TFeedbackResult>` to create a register (that acts as a factory) for a handler
- Tests for everything above mentioned to reduce the amount of them needed per individual component (as they share the logic anyway)
- Base Tests classes to reduce boiler-plate needed. Refer to `UIComponentsShouldBase`
- `UITextureUpdater`, `UIFontUpdater` to handle corresponding promises independently on elements logic

It allows us to focus on individual unique logic per elements, leaving the repetitive boilerplate behind the scene.

### UIInput

- Additional helper class `TextFieldPlaceholder` to handle placeholder. This logic does not exist in the current Unity version.

### UIDropdown

- Contains additional logic for a default value as it is not supported out of box if the element is already instantiated

### UIBackground

- New `DCLImage`
   - Wraps up scaling modes that are not supported out of box
   - Stretch mode supports custom UVs so it is possible to render scewed/rotated images without limitations
   - Based on custom mesh generation
   - Reuses existing logic for several modes
   - Abstracted from the `Proto` domain, belongs to `UIElements` domain
   - Draws on a parent elements (to avoid an extra element and act as a background)

### Styling
Unity applies default styles for visual elements, we don't provide sufficient options for content creators to cover the elements needs.
Right now the elements will be functional but rendering of them is partly customizable and will be based on default styles.
As an example you may think of `UIDropdown`'s items in the list for which there are no fields in proto to customize background, frames, etc.

**TODO:** Create an ADR to approach it wisely and properly

### Foreground elements ambiguity 
From the logical perspective it makes no sense to add several foreground elements (such as `UIInput`, `UiText`, `UIDropdown`) to a single entity. It should be forbidden from the SDK side.

### Code smells and propositions

- A separate visual element corresponding to `UITransform` is redundant when there is a `UIBackground` component attached, or any other foreground component without `UIBackground`:
   - Each `VisualElement` is a fully fledged transform so it can act as a replacement for the `UITransform`'s element.
   - We can think of a postponed creation based on all components of an entity

## How to test the changes?

1. Every possible combination of parameters of UI components should be checked. Functionality [described here](https://github.com/decentraland/adr/issues/125) should be covered. From the ADR you may see which fields are optional (and thus can be omitted in a component definition) and which are not.
2. Refer to schemes
3. Create/checkout a scene to play with. You can use `sdk7-scene-template` as a reference
4. NPM Install the latest revision of `jsd-sdk-toolchain`  `pointer-event` branch
5. Launch a test scene and modify the URL (in a browser) to include `ws` like this: `http://127.0.0.1:8000/?position=-1%2C-2&SCENE_DEBUG_PANEL=&ENABLE_ECS7=&ws=wss%3A%2F%2Flocalhost%3A7666%2Fdcl&realm=LocalPreview`
6. In Unity Editor set `Base Url Mode` to `LOCAL_HOST`. Now it should connect to that test scene
7. Introduce as many changes to UI Entities as needed to test, they will reloaded on fly

### Schemes
```
uiDropdown={{
          acceptEmpty: true,
		  emptyLabel: 'default value',
		  options: ['option1', 'option2', 'option3'],
		  disabled: false,
		  color: Color4.Blue(),
		  textAlign: TextAlignMode.TAM_BOTTOM_CENTER,
		  fontSize: 18,
		  onChange: (value: number) => {console.log(value)}
        }}
```
```
uiInput={{
placeholder: 'Boedo its carnaval',
      disabled: false,
      color: Color4.Red(),
      placeholderColor: Color4.Blue(),
      textAlign: TextAlignMode.TAM_BOTTOM_CENTER,
      font: Font.F_SANS_SERIF,
      fontSize: 14,
      onChange: (value: string) => {console.log(value)}
}}
```
```
uiBackground={{
		color: Color4.fromInts(231,255,255,255),
		textureMode: BackgroundTextureMode.STRETCH,
		texture: {
			src: 'images/nature.png',
			wrapMode: TextureWrapMode.TWM_CLAMP,
			filterMode: TextureFilterMode.TFM_BILINEAR
		},
		textureSlices: {
			top: 0.47,
			left: 0.59,
			right: 0.47,
			bottom: 0.38
		},
		uvs: [0.2, 0.2, 0.2, 0.5, 0.55, 0.5, 0.55, 0.2]
	  }}
      />
```

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md

### CODE REVIEW

File: UNITY-~1/Assets/Scripts/MAINSC~1/DCL/CONTRO~1/EXPLOR~1/Scripts/Sections/PLACES~1/SUBSEC~1/HIGHLI~1/HIGHLI~1/HIGHLI~1.CS
Comment (menduz) [inline UNITY-~1/Assets/Scripts/MAINSC~1/DCL/CONTRO~1/EXPLOR~1/Scripts/Sections/PLACES~1/SUBSEC~1/HIGHLI~1/HIGHLI~1/HIGHLI~1.CS:1]: This filename can't be ok. Can you double check it?

File: protocol-gen/package.json
Comment (menduz) [inline protocol-gen/package.json:18]: Before merging we should point to a published version

File: unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/UIAbstractElements/UIElementRegisterBase.cs
Comment (menduz) [inline unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/UIAbstractElements/UIElementRegisterBase.cs:8]: Can you add comments about the objective of this classes?

File: unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/UIComponentsUtils/UIEventSubscription.cs
Comment (menduz) [inline unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/UIComponentsUtils/UIEventSubscription.cs:48]: Should this be analog 0?

File: unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/UIComponentsUtils/UIFontUpdater.cs
Comment (menduz) [inline unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/UIComponentsUtils/UIFontUpdater.cs:5]: <3

**Review by menduz — COMMENTED** [review]
Awesome job! I will review in the playground and complement this initial comments if needed

File: unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/UIComponentsUtils/UIEventSubscription.cs
Comment (mikhail-dcl) [reply unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/UIComponentsUtils/UIEventSubscription.cs:48]: I believe it's irrelevant for UI, so I can just omit this field

File: unity-renderer/Assets/DCLPlugins/ECS7/UIElements/Image/IUITextureConsumer.cs
Comment (Kinerius) [inline unity-renderer/Assets/DCLPlugins/ECS7/UIElements/Image/IUITextureConsumer.cs:1]: you missed this one 😆

**Review by Kinerius — APPROVED** [review]
LGTM!

**Review by menduz — APPROVED** [review]
Awesome job! Background is working as expected, I've created other visual and functional issues to attack later on. This should be merged as soon as possible to unblock other PRs that depend on the API but not on the correct behavior.

https://github.com/decentraland/sdk/issues/549
https://github.com/decentraland/sdk/issues/547
https://github.com/decentraland/sdk/issues/552
https://github.com/decentraland/sdk/issues/551

File: unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/PointerEvents/UIPointerEventsHandler.cs
Comment (pbosio) [inline unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/PointerEvents/UIPointerEventsHandler.cs:12]: we are, strongly, trying to avoid any kind on inheritance on the sdk7 code, specially regarding components handlers

File: unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/UIAbstractElements/UIElementHandlerBase.cs
Comment (pbosio) [inline unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/UIAbstractElements/UIElementHandlerBase.cs:12]: all this logic could be replaced by extending `IInternalECSComponent<InternalUiContainer>` like we have for other components, for example https://github.com/decentraland/unity-renderer/blob/dev/unity-renderer/Assets/DCLPlugins/ECS7/InternalECSComponents/Interfaces/Extensions/InternalCollidersExtensions.cs
instead of adding this abstract class

File: unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/UIComponentsUtils/UIEventsSubscriptions.cs
Comment (pbosio) [inline unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/UIComponentsUtils/UIEventsSubscriptions.cs:13]: could we replace this for a the good old `for` instead of using `foreach`

File: unity-renderer/Assets/DCLPlugins/ECS7/InternalECSComponents/Interfaces/InternalECSComponentModels.cs
Comment (pbosio) [inline unity-renderer/Assets/DCLPlugins/ECS7/InternalECSComponents/Interfaces/InternalECSComponentModels.cs:68]: why are we adding this?

**Review by pbosio — APPROVED** [review]
Awesome work! 
I left a couple of comments, but can be tackled after this PR is merged if we are in a hurry to merge it.
There are 4 `.asset` modifie, is that intended?

**Pull request merged**
